lib/rest/Twilio.js

```diff
@@ -249,18 +253,36 @@ Twilio.prototype.request = function request(opts) {
     headers.Accept = 'application/json';
   }
 
-  var uri = opts.uri;
-  if (this.region) {
-    var parts = _.split(uri, '.');
-
-    if (parts.length > 1 && !_.isEqual(parts[1], this.region)) {
-      uri = _.join(_.concat([parts[0], this.region], _.slice(parts, 1)), '.');
+  const domain = 'twilio.com';
+  var product;
+  var region;
+  var edgeOrRegion;
+  var uri = new url.URL(opts.uri);
+  var prefix = uri.hostname.replace('.'.concat(domain), '');
+  [product, edgeOrRegion, region] = prefix.split('.');
+  if (edgeOrRegion && region) { // https://product.edge.region.subdomain.domain
+    if (this.edge) {
+      uri.hostname = [product, this.edge || edgeOrRegion, this.region || region, domain].join('.');
+    } else if (this.region) {
+      uri.hostname = [product, this.edge || edgeOrRegion, this.region, domain].join('.');
+    }
+  } else if (edgeOrRegion && !region) { // https://product.region.subdomain.domain
+    if (this.edge) {
+      uri.hostname = [product, this.edge || edgeOrRegion, this.region || region, domain].join('.');
+    } else if (this.region) {
+      uri.hostname = [product, this.region || edgeOrRegion, domain].join('.');
+    }
+  } else if (!edgeOrRegion && !region) { // https://product.subdomain.domain
+    if (this.edge) {
+      uri.hostname = [product, this.edge, this.region || defaultRegion, domain].join('.');
+    } else if (this.region) {
+      uri.hostname = [product, this.region, domain].join('.');
```

Comment:
Still overly complex. What about something like this:
```
function getHostname(uri, targetEdge, targetRegion) {
  const domain = 'twilio.com';
  const defaultRegion = 'us1';

  targetRegion = targetRegion || (targetEdge && defaultRegion);

  if (!targetRegion) {
    return uri;
  }

  const prefix = uri.split('.' + domain)[0];
  let [product, edge, region] = prefix.split('.');

  if (!region || targetEdge) {
    edge = targetEdge;
  }

  region = targetRegion;

  return [product, edge, region, domain].filter(part => part).join('.');
}
```

Reply:
Updated, thanks!